### PR TITLE
Prevent credentials edition for sites with no purchases

### PR DESCRIPTION
**Pull request by monsieur-z**
### Changes proposed in this Pull Request

Sites that never had any purchased plan or product can't currently have user credentials be saved. See 1164141197617539-as-1199981665376903. This PR improves the user experience a bit by showing an empty state if the site has no purchased plans nor products.

#### Changing site

When making a request to fetch a site purchases, information about the state of the request (ongoing and completed) is stored in the state independently of the site (respectively `state.purchases.isFetchingSitePurchases` and `state.purchases.hasLoadedSitePurchasesFromServer`). This is limiting in our case, since the loading condition in `HasSitePurchasesSwitch` will keep returning false after switching site.

To prevent this, we could isolate the site specific request information in the state (e.g. `state.purchases.requests[siteId].isFetching`), but that's too big of a change, at least for the scope of this PR.

A simpler solution is to detect the site switch in `HasSitePurchasesSwitch`, and to reset the `isFetchingSitePurchases` and `hasLoadedSitePurchasesFromServer` values.

#### Testing

Unit tests were added for the newly created components. Most of them doesn't include any logic, so I've just included snapshots to make sure further changes are intentional.

### Testing instructions

- Download the PR and run cloud
- Create a new self-hosted Jetpack site, and connect it (don't purchase any plan or product)
- Visit `/settings/:site`
- Notice the loading placeholder while purchases are being fetched, then the empty state (see video below)
- Check that the button links to `jetpack.com/pricing`
- Switch section back and forth, and check that you see the empty state instantly
- Select another site that has a purchased product or plan
- Check that you can add your credentials
- Remove any purchased product or plan
- Make sure you can still remove or edit the credentials

### Screenshots

https://user-images.githubusercontent.com/1620183/111787478-5905f780-8895-11eb-833c-9448796ed3ea.mov

### CODE REVIEW

**Comment by monsieur-z** [comment]
Any feedback for the notice copy is welcome.

**Review by rcanepa — APPROVED** [review]
Code changes look good and the form behaves as expected. 

LGTM!

**Comment by rcanepa** [comment]
@monsieur-z, let me know if you need a second review.

File: client/components/has-site-purchases-switch/index.tsx
Comment (monsieur-z) [inline client/components/has-site-purchases-switch/index.tsx:47]: Will reset loading information. See implementation notes.

File: client/components/has-site-purchases-switch/test/index.jsx
Comment (monsieur-z) [inline client/components/has-site-purchases-switch/test/index.jsx:18]: It was easier here to mock the query component rather than mocking the endpoint.

File: client/jetpack-cloud/sections/settings/controller.tsx
Comment (monsieur-z) [inline client/jetpack-cloud/sections/settings/controller.tsx:29]: Site purchases now need to be fetched before rendering anything.

File: client/jetpack-cloud/sections/settings/loading/style.scss
Comment (monsieur-z) [inline client/jetpack-cloud/sections/settings/loading/style.scss:10]: Prevent illustration from "jumping" after it has finished loading.

**Comment by monsieur-z** [comment]
> @monsieur-z, let me know if you need a second review.

Thanks @rcanepa. I've updated the instructions, since changes are quite significative.

**Comment by mattgawarecki** [comment]
*(EDIT: See my next comment)* ~Everything~ Almost everything you mention in the PR description works for me, and the unit tests are clear. I know snapshots aren't ideal, but given there's no logic, I think this is much better than not having them at all.

I was _just_ about to approve, but I noticed on mobile layouts there's no page title in the top bar and no way for me to switch sites (see screenshots). Can you add those in?

**With purchases (left) / without purchases (right)**

image image

**Comment by mattgawarecki** [comment]
Actually, I just went back to confirm I could delete credentials from a site with no purchases but with credentials having been saved beforehand. It looks like maybe not? I'm getting the "empty content" page when I try.

**Comment by monsieur-z** [comment]
> I was _just_ about to approve, but I noticed on mobile layouts there's no page title in the top bar and no way for me to switch sites (see screenshots). Can you add those in?

I'm not seeing the Manage Sites dropdown at any viewport, @mattgawarecki. Do you know what are the conditions to show it?

**Comment by monsieur-z** [comment]
> Actually, I just went back to confirm I could delete credentials from a site with no purchases but with credentials having been saved beforehand. It looks like maybe not? I'm getting the "empty content" page when I try.

You're right, I forgot to carry over that part after I changed the implementation. Updated.

File: client/jetpack-cloud/sections/settings/controller.tsx
Comment (monsieur-z) [inline client/jetpack-cloud/sections/settings/controller.tsx:30]: The site credentials and purchases requests should ideally be done concurrently, but I favoured having 2 separate switches here for clarity. It seems that credentials data are often already in the state at this point, so the gain of doing that seems to be minimal.

File: client/jetpack-cloud/sections/settings/has-site-credentials-switch.tsx
Comment (monsieur-z) [inline client/jetpack-cloud/sections/settings/has-site-credentials-switch.tsx:35]: We want to prevent further calls to the credentials endpoint to trigger the loading state again.

**Review by mattgawarecki — APPROVED** [review]
Everything tested well for me this time. Nice work! 👍 

Sidenote: I think we'll eventually want to come back and make some copy and/or design changes to reflect this "credentials exist but no active purchases" state, since we still mention Jetpack being able to perform restores and scan for threats.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5647767

Hi @monsieur-z, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Upgrade for quick restores`
- `Server credentials are used to restore your site and fix any vulnerabilities`

Thank you in advance!

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5647767).

**Pull request merged**
